### ISSUE TITLE
Support concrete default for `impl` argument.

### ISSUE THREAD

**Issue by zakstucke** [elastio/bon#168]
Great library! 

It would be really powerful to make it possible to provide defaults for generic parameters, the below example fails to compile:

```rust
#[builder]
fn foo(#[builder(default = std::iter::empty())] arg: impl IntoIterator<Item = String>) {
    drop(arg);
}
```
```
213 |     #[builder]
    |     ---------- expected this type parameter
214 |     fn foo(#[builder(default = std::iter::empty())] arg: impl IntoIterator<Item = String>) {
    |                                ^^^^^^^^^^^^^^^^^^ expected type parameter `I1`, found `Empty<_>`
```

### A note for the community from the maintainers

Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to help the maintainers with prioritizing it. You may add a comment describing your real use case related to this issue for us to better understand the problem domain.

**Comment by Veetaha** [comment]
Hi, thank you for creating the issue! Unfortunately, this isn't possible with the current design of the builder struct's type signature. The builder captures the generic parameters eagerly.

There is a [cool article](https://geo-ant.github.io/blog/2024/rust-rethinking-builders-lazy-generics/) that explores the design where generic parameters are resolved lazily, but that approach is a huge rabbit hole of complexity and probably will lead to unacceptable compile times.

Additionally, I think using `#[builder(default)]` with generic parameters (including `impl Trait`) is an antipattern due to the current design of "eager generic parameters". I'm planing to document that antipattern for the next release. 

In short the problem is the following:

```rust
#[bon::builder]
fn example(#[builder(default)] value: impl Default) {}

let _ = example().call();
```

This code doesn't compile with the error because the compiler can't infer the type of the omitted `impl Default` parameter:

```
type annotations needed
cannot satisfy `_: std::default::Default`
```

Instead, once the next version of bon (#156) is released, you could write this code like this:

```rust
#[bon::builder]
    fn foo(
        #[builder(default = vec![], with = FromIterator::from_iter)]
        arg: Vec<String>
    ) {
        drop(arg);
    }

    let _ = foo().call();
```
The new `#[builder(with = ...)]` annotation allows you to move the generic parameters to the setter, removing it from the builder struct. So the setter still accepts `impl IntoIterator<Item = String>`, but in the function you already receive a concrete collection type.

Let me know if this makes sense to you and you understand the footguns with the generic types inference here. I'll need to document this and want to make sure the wording is clear.

**Comment by Veetaha** [comment]
I see. Well, another way to work around this is to wrap the value in `Option<>` and handle it internally:

```rust
fn foo(arg: Option<impl MyTrait>) {
    if let Some(arg) = arg {
        // ...
    } else {
        // ...
    }
}
```

Although... there is still a problem with infering the generic type if you don't pass a value to a setter, so this doesn't compile:

```rust
foo().call();
```

instead you'd need to write this:

```rust
foo::<TraitImplType>().call()
```

Unfortunatelly, this is the limitation of the language where you can't assign default values for generic type parameters in functions (although that's possible to do with generic struct/enum/union)

**Comment by zakstucke** [comment]
Yep it's not possible to solve without bon/builders. I was imagining with bon it could work like:

```rust

trait MyTrait;

impl MyTrait for ();

impl MyTrait for usize;

#[builder]
fn foo(arg: impl MyTrait) {}

foo() // Builder<()>
    .call() // foo::<()>

foo() // Builder<()>
    .arg(10) // Builder<usize>
    .call() // foo::<usize>
```

but I don't claim to know the internals or that this would work. 

Feel free to close this issue if you're sure it won't be possible to action :)

**Comment by Veetaha** [comment]
Interesting idea... Although in general case, and after desugaring this should work with named generic types and the user needs to clarify the exact type of thee default value:

```rust
#[builder]
fn foo<T: MyTrait = ()>(#[builder(default = ())] arg: T) {
}
```

I think `syn` accepts default values for generic types in function position like that, so this may be actually workable with this syntax where the user has to uses the verbose named generic type parameter syntax, and has to specify the default value for the generic type for which the default expression is written.

The tough part is that the setter needs to be generated like this:

```rust
impl<T, State> FooBuilder<T, State> {
     fn arg<U>(value: U) -> FooBuilder<U, SetArg<State>> { /**/ }
}
```
This is where the macro needs to figure out the proper generics for the `arg`. For example, when there are many generic parameters, the macro needs to only add generics that are used in the type of the `arg` and ignore all the rest. 

This approach is definitely workable in simple cases, but needs some more prior design to prove that it can work in a general case with complex `T::Assoc`, `where ...` clauses and other stuff.

**Comment by Veetaha** [comment]
A problem with this approach.. How should the finishing function be written?

```rust
impl<T, State> FooBuilder<T, State> {
    fn call(self) {
        let arg: T = self.arg.unwrap_or(|| default_expression);
        __org_foo(arg)
    }
}
```

This suffers from the same problem where we need to cast `default_expression` to `T`. Maybe this can be achievable with some unsafe `transmute` or some other type system cool trick.. Need to think about this more

**Comment by zakstucke** [comment]
> How should the finishing function be written?

By not having `arg` as `Option<T>` but instead `T`, or `FnOnce(T)` for performance. The value shouldn't be set to default at the end, but at the beginning with this approach.

**Comment by Veetaha** [comment]
> The value shouldn't be set to default at the end, but at the beginning

That's unacceptable. Imagine the default value performs some complex computation or allocation. Calculating it eagerly would require paying the cost of that even if the default value is overridden.

Also, the default expression today guarantees access to already set members in scope. From [the docs](https://bon-rs.com/reference/builder#default):

> You can also use the values of other members by referencing their names in the default expression. All members are initialized in the order of their declaration. It means only those members that are declared earlier (higher) in the code are available to the default expression.

Storing `FnOnce(T)` in the builder's fields is not possible because how would you then set an override for the value? Also `FnOnce` requires yet another generic parameter on the builder's struct for the type of the function itself.

**Comment by Veetaha** [comment]
I could make it work like this with a `transmute` (although it doesn't compile because the names of internal builder fields are generated with random words in them to ensure their privacy, but that wouldn't be a problem if this code was generated instead).

```rust
#[bon::builder(
    builder_type = FooBuilder,
    finish_fn(vis = "", name = call_internal),
)]
fn foo_internal(#[builder(setters(name = arg_internal))] arg: impl IntoIterator<Item = String>) {}

fn foo() -> FooBuilder<std::iter::Empty<String>> {
    foo_internal()
}

impl<T: IntoIterator<Item = String>, State: foo_builder::State> FooBuilder<T, State> {
    fn arg<U: IntoIterator<Item = String>>(self, value: U) -> FooBuilder<U, State>
    where
        State::Arg: foo_builder::IsUnset,
    {
        FooBuilder {
            __private_twilight_phantom: std::marker::PhantomData,
            __private_twilight_named_members: (Some(value),),
        }
    }

    fn call(self) {
        let arg: T = self.__private_twilight_named_members.0.unwrap_or_else(|| {
            let default: std::iter::Empty<String> = std::iter::empty();
            unsafe { std::mem::transmute(default) }
        });

        __orig_foo_internal(arg)
    }
}

fn main() {
    foo().call();

    foo().arg(["foo".to_owned()]).call();
}
```

But then I completely forgot about the `maybe_` version of the setter that accepts an `Option<T>`. I don't see how it could be written:

```rust
let _: Builder<??, _> = foo().maybe_arg(Some(["str".to_owned()]));
```

What would be the type `??`? Is it `[1; String]` or is it `std::iter::Empty<String>`? The problem is that the type of that generic parameter strictly depends on wether a `Some` or `None` was passed. So such setter wouldn't be possible to write

**Comment by Veetaha** [comment]
There is a lot to research and experiment with here. I don't see an obvious design that would fit into the existing model without compromising some existing features, but maybe there is such a design that satisfies all requirements. I'd leave this issue open until a comprehensive design appears that takes into account all the edge cases. The first thing to do is to have a code snippet of the generated code like I did above that makes it obvious how generation of such code could be automated in a macro and how it would interact with other existing configs of the macro.

I'll try to think about this more once after I close the current higher priority items
